Python: Refocus hosting channels ADR on protocol helpers (#6837)
microsoft/agent-framework: docs/decisions/0027-hosting-channels.md

  response/event payloads;
- protocol-specific session-key helper functions when useful, such as `telegram_session_key(chat_id)`;
- protocol-specific typing/update event helpers where the protocol has a native concept.

Application or web-framework code owns:

- HTTP route declaration and route grouping;
- dependency injection;
- authentication and authorization;
- middleware;
- background tasks and webhook acknowledgement policy;
- native protocol SDK clients and outbound calls;
- command registration and command dispatch;
- request/response status codes and framework-specific error handling.

`AgentFrameworkHost`, if provided, is limited to shared execution state:

- the FastAPI/Starlette app object or equivalent app handle when useful;
- one hostable target (`SupportsAgentRun` agent-compatible object or a `Workflow`);
- a `SessionStore` instance or factory;
- optional workflow checkpoint execution state.

It is **not** a channel registry and does not own route contribution, protocol dispatch, command projection, or native
SDK calls.

### Helper naming

Helpers should be protocol-specific, not generic. Prefer:

- `responses_to_af(...)`

> - `response_from_af(...)`

Author: Addressed in 05da96ea2: the helper names now use protocol-specific run terminology (`responses_to_run(...)` / `responses_from_run(...)`) instead of the older `*_af` naming, so the Responses helper is no longer generic or inconsistent.

- `ChannelStreamUpdateHook`
- host-owned route aggregation
- host-owned lifecycle aggregation
- host-owned command projection
- host-owned protocol dispatch

Apps can implement equivalent local behavior when needed using their web framework and protocol SDK. Shared behavior
that proves broadly useful can be reconsidered after concrete samples justify it.

## Non-goals for v1

The following remain outside the v1 contract:

- cross-channel identity linking (`IdentityLinker`, `local_identity_link`, or `agent-framework-hosting-entra`);
- identity allowlists or authorization policy (`IdentityAllowlist`, `AuthPolicy`);
- response routing beyond the originating protocol (`ResponseTarget`, active channel, specific linked channel,
  `all_linked`);
- push or payload codecs (`ChannelPush`, `ChannelPushCodec`);
- background/continuation delivery;
- durable task runners (`DurableTaskRunner`, `InProcessTaskRunner`);
- retry/replay policy (`RetryPolicy`);
- fan-out, multicast, or all-linked delivery;
- confidentiality tiers and `LinkPolicy`;
- a host-level multi-agent router.

These areas are follow-up enhancements covered by [ADR-0028](0028-hosting-linking-multicast-enhancements.md). They are

> not prerequisites for shipping or using the v1 protocol-helper surface.

Author: Addressed in 05da96ea2: ADR-0027 now explicitly notes that ADR-0028 was written against the earlier host/channel framing and must be revised to align with the protocol-helper/execution-state boundary before implementation.

## Consequences

Positive:

- The released surface is smaller and easier to inspect: helpers plus state, not a channel framework.
- Protocol helpers can be used from FastAPI, Starlette, Azure Functions, Django, CLI tools, tests, or native SDK webhook
  handlers.
- App authors can use the authentication, dependency injection, lifecycle, and background-task tools they already know.
- Session continuity stays explicit and debuggable.
- Workflow checkpointing can still be centralized if needed without making protocol packages own routing.

Negative:

- Multi-protocol samples include explicit route/client code.
- Apps that want a batteries-included ASGI host must write or depend on an app-specific wrapper.
- Existing unreleased code and docs that mention channels, contribution, or hooks must be revised before release.

## Validation Gates

Before this ADR is considered implemented:

> - A Responses sample uses normal FastAPI route code plus `responses_to_af(...)`, `response_from_af(...)`, and

Author: Addressed in 05da96ea2: the validation gate now refers to `responses_to_run(...)` / `responses_from_run(...)`, matching the helper naming section.

protocol data, while app/native SDK code performs acknowledgements, sends/edits messages, resolves protected file URLs,
applies rate limits, and registers handlers.

### Session continuity

Session continuity remains explicit. Run parsing and isolation/session id selection are separate operations because
isolation can come from more than one source:

- protocol input, such as OpenAI Responses `previous_response_id`, a Telegram chat id, or an Activity conversation id;
- running environment, such as Foundry Hosted Agents user/chat isolation context;
- app-specific trusted middleware or route state.

The app chooses which helper to call for that route and deployment. For example:

- `responses_session_id(body)` from `agent-framework-hosting-responses`, which can return either a `resp_*` previous
  response id or a `conv_*` conversation id when present;
- `telegram_session_id(update)` from `agent-framework-hosting-telegram`, which can choose the chat, user, thread, or
  other Telegram-native partitioning logic for that helper;
- `activity_session_id(activity)`, `discord_session_id(interaction_or_message)`, or
  `a2a_session_id(request_context)` from their respective protocol packages;

> - `foundry_user_isolation_key()` or `foundry_chat_isolation_key()` from `agent-framework-foundry-hosting`.

Reviewer: With the Validation Gates section removed, is anything still requiring these helpers to fail closed outside Foundry? The previous revision explicitly gated on tests proving raw isolation headers are ignored in the non-Foundry case. This revision keeps the helpers as the trusted platform-provided key source but drops that requirement, so an implementation that reads isolation headers verbatim would let a caller spoof another user's isolation key when the app runs outside Foundry.

Author: Addressed in ca599835f: ADR-0027 now explicitly says platform-provided isolation helpers must fail closed outside their trusted hosting environment. It also states Foundry-specific helpers may read values established by Foundry hosting middleware, but must not treat raw request headers as trusted Foundry isolation outside Foundry, and implementations must test that non-Foundry requests do not accept spoofable isolation headers.

externally supplied key before using it.

### Workflow checkpoints

Workflow checkpointing is execution state, not protocol state. `WorkflowState` pairs a workflow target with a
`CheckpointStore` (`session_id -> CheckpointStorage`) and provides `get_or_create_checkpoint_storage(...)` for first use.
The store itself remains plain storage and does not decide which checkpoint to resume.

For workflow targets, app code adapts the protocol helper output into the workflow's expected input and invokes the
workflow through the state object's target:

```python
storage = await state.get_or_create_checkpoint_storage(session_id)
target = await state.get_target()
result = await target.run(message=workflow_input, checkpoint_storage=storage)
await state.set_checkpoint_storage(session_id, storage)
```

If a route wants to resume from a prior checkpoint, it explicitly chooses the checkpoint and passes it to
`workflow.run(...)`:

```python
storage = await state.get_or_create_checkpoint_storage(session_id)
target = await state.get_target()
latest = await storage.get_latest(workflow_name=target.name)

> result = await target.run(

Reviewer: Should this resume snippet drop `message` when `latest` exists? `Workflow.run` validates that `message` and `checkpoint_id` are mutually exclusive (`_workflows/_workflow.py` raises "Cannot provide both 'message' and 'checkpoint_id'"), so as written this raises `ValueError` exactly when a checkpoint is found. Since the appendix imports the real `Workflow` and `CheckpointStorage`, this reads as a contract against the existing API rather than a future one.

Author: Addressed in ca599835f: the workflow resume snippet now branches. If no checkpoint id is available it calls run(message=..., checkpoint_storage=...). If a checkpoint id exists it calls run(checkpoint_id=..., checkpoint_storage=...) and does not pass message, matching the current Workflow.run validation.

Negative:

- Multi-protocol samples include explicit route/client code.
- Apps that want a batteries-included ASGI app must write or depend on an app-specific wrapper.
- Existing unreleased code and docs that mention channels, contribution, or hooks must be revised before release.

## More Information

- Follow-up linking and multicast ADR: [ADR-0028](0028-hosting-linking-multicast-enhancements.md). That ADR still uses
  some earlier host/channel terminology and must be aligned before implementation work starts.

## Appendix: Developer experience sketch

### Optional execution state

`AgentState` and `WorkflowState` stay small: they are target-specific state holders, not app hosts.

```python
from typing import Protocol

from agent_framework import AgentSession, CheckpointStorage, SupportsAgentRun, Workflow

class SupportsBuild(Protocol):
    def build(self) -> Workflow: ...

class SessionStore:
    async def get(self, session_id: str) -> AgentSession | None: ...
    async def set(self, session_id: str, session: AgentSession) -> None: ...
    async def delete(self, session_id: str) -> None: ...

> class CheckpointStore:

Reviewer: Can a non-in-memory `CheckpointStore` actually implement `set`? `CheckpointStorage` is a protocol whose implementations hold live clients (the Cosmos-backed one holds a Cosmos client), so a Redis or Cosmos backed store cannot persist a live storage instance by value, yet the async get/set/delete surface and the "plain async storage" wording at line 96 signal out-of-process backends are the intent. Would mapping `session_id` to storage config/ids, or using one partitioned `CheckpointStorage`, keep multi-replica session continuity implementable?

Author: Addressed in ca599835f: the ADR no longer frames durable CheckpointStore as storing live CheckpointStorage clients by value. It now distinguishes local/in-memory session_id -> CheckpointStorage state from durable/multi-replica session_id -> checkpoint_id (or equivalent cursor/config) state, and notes that WorkflowRunResult/events do not emit checkpoint ids by default; app/state code that owns storage can observe the latest id via storage.get_latest(workflow_name=target.name) after a run.

from fastapi import Body, FastAPI, Header, HTTPException
from fastapi.responses import JSONResponse

app = FastAPI()
agent = Agent(
    client=OpenAIChatClient(),
    name="Assistant",
    instructions="Be concise and helpful.",
)
state = AgentState(agent)

@app.post("/responses")
async def responses(body: dict = Body(...), x_api_key: str | None = Header(default=None)) -> JSONResponse:
    if x_api_key != os.environ["RESPONSES_API_KEY"]:
        raise HTTPException(status_code=401, detail="bad api key")

    # parse the request body into a set of AF objects
    run = responses_to_run(body)
    # get the session id from the body
    # can be a resp_* for previous_response_id or a conv_* for a conversation
    session_id = responses_session_id(body)
    # create a new response_id for this run
    response_id = create_response_id()

    # in this space, the developer can make any adjustments to the request, i.e.:
    run["options"]["store"] = False
    run["options"].pop("model", None)

    # load the session (or create a new one) - this is optional

>     session = await state.get_or_create_session(session_id or response_id)

Reviewer: Should this sketch show, or at least note, an ownership check on the caller-supplied id? After the shared api key check, any caller can send another conversation's `resp_*` or `conv_*` id and inherit that `AgentSession` with its full history, which is exactly the pattern line 187 says apps must authorize first. The Django sketch repeats the same shape, and these examples are what samples and implementations will copy. Even a `# verify this caller owns session_id before loading` comment would keep the sketches aligned with the trust note.

Author: Addressed in ca599835f: both the FastAPI and Django sketches now include an explicit comment before loading the session that the app must verify this caller owns the supplied session id; API-key auth alone does not prove ownership of a caller-supplied resp_* or conv_* id.

from agent_framework import Agent
from agent_framework.openai import OpenAIChatClient
from agent_framework_hosting import AgentState  # pyright: ignore[reportAttributeAccessIssue]
from agent_framework_hosting_responses import create_response_id, responses_from_run, responses_session_id, responses_to_run  # pyright: ignore[reportAttributeAccessIssue]
from asgiref.sync import async_to_sync
from django.http import HttpRequest, HttpResponseBadRequest, HttpResponseForbidden, JsonResponse
from django.views import View

agent = Agent(
    client=OpenAIChatClient(),
    name="Assistant",
    instructions="Be concise and helpful.",
)
state = AgentState(agent)

class ResponsesView(View):
    def post(self, request: HttpRequest) -> JsonResponse:
        if request.headers.get("x-api-key") != os.environ["RESPONSES_API_KEY"]:
            return HttpResponseForbidden("bad api key")

        try:
            body = json.loads(request.body)
        except json.JSONDecodeError:
            return HttpResponseBadRequest("invalid json")

        run = responses_to_run(body)
        session_id = responses_session_id(body)
        response_id = create_response_id()
        options = run["options"]

>         session = async_to_sync(state.get_or_create_session)(session_id or response_id)

Reviewer: Has this Django shape been tried against a second request? Each `async_to_sync` call in a sync view runs on a fresh event loop, but the module-level `OpenAIChatClient` caches an `AsyncOpenAI` client whose pooled httpx connections bind to the loop of the first `target.run`, so later requests typically fail with "Event loop is closed". Could the sketch use an async view (`async def post`) so the whole request runs on one live loop?

Author: Addressed in ca599835f: the Django sketch is now an async view (`async def post`) and uses await for state.get_or_create_session, state.get_target, target.run, and state.set_session. The example no longer uses async_to_sync around a module-level async client.

---
status: accepted
contact: eavanvalkenburg
date: 2026-06-30
deciders: eavanvalkenburg
---

# Python protocol helpers and optional execution state

## Context and Problem Statement

Agent Framework needs to help applications expose agents and workflows over external protocols such as OpenAI
Responses, Telegram, Activity Protocol, and future transports.

> The first version of this ADR chose a host/channel model: channel packages contributed routes, middleware, commands,

Reviewer 2: nit: perhaps do not mention the first version at all if it won't exist in the code base.

`CheckpointStorage` abstraction directly; app/state code may keep a small cursor (`session_id -> checkpoint_id`) when it
needs to resume a workflow for a session.

These objects are **not** app objects, channel registries, or route owners. They do not own FastAPI/Starlette setup,
route contribution, protocol dispatch, command projection, or native SDK calls.

### Helper naming and families

Helpers should be protocol-specific, not generic. Avoid a generic `protocol_to_run(...)` name in public samples because it
hides the protocol-specific contract behind a second abstraction.

Protocol packages should consider these helper families. This table is a set of examples, not a required protocol or
checklist. Not every protocol needs every helper, but when a protocol has the concept the naming should stay consistent:

| Helper family | Shape | Purpose |
| --- | --- | --- |
| Run conversion | `<protocol>_to_run(...)` | Convert one protocol-native call/update/request into `Agent.run` or `Workflow.run` values. |
| Final rendering | `<protocol>_from_run(...)` | Convert a final `AgentResponse` / workflow result into protocol-native response payloads or operations. |

> | Stream rendering | `<protocol>_stream_from_run(...)` | Convert `ResponseStream` / workflow updates into protocol-native events or operations. |

Reviewer 2: nit: `<protocol>_from_streaming_run(...)`

Author: good point

A `SessionStore` stores `session_id -> AgentSession`, but it does not create sessions. `AgentState` resolves the agent
target and creates the session on first use:

For agent targets:

```python
session = await state.get_or_create_session(session_id)
target = await state.get_target()
result = await target.run(messages, session=session, options=options)
```

If the protocol mints a new continuation id as part of the response being created (for example, OpenAI Responses
`resp_*` ids), store the **post-run** session explicitly under that new id:

```python
session = await state.get_or_create_session(previous_response_id)
target = await state.get_target()
result = await target.run(messages, session=session, options=options)
await state.set_session(response_id, session)
```

`agent.run(...)` may update the session object (for example, with service continuation state), so the explicit store call
belongs after the run, not before it.

The session id is a partition key, not proof of identity. App or platform code must authenticate and authorize any
externally supplied key before using it.

### Workflow checkpoints

> Workflow checkpointing is execution state, not protocol state. `WorkflowState` pairs a workflow target with checkpoint

Reviewer 2: Is a "workflow target" a workflow instance? How does this target survive across sessions?

Author: See line 89, it refers to: `- ``WorkflowState``: one ``Workflow``, ``WorkflowBuilder``-shaped builder, orchestration builder, or workflow factory;`

In other words, depending on the workflow, hosting api, use case and other things, you can choose to build a workflow anew for each run, or reuse a existing one, the WorkflowState has a target attribute that can be set (WorkflowState itself is optional), if needed augmented with a small lookup between i.e. response_id and checkpoint_id to continue when needed (where response_id could also be chat_id for telegram, or whatever identifier there is for a particular channel)

Author: And also note (and I added this as well) that the developer chooses where they intend to run, and whether that is a persistent server, like a Container App, or a transient thing like AzFunctions, because that obviously has major implications for how to design storage and which is why I have set this up with either a concrete instance of a agent or a callable, so that both options are present.

---
status: accepted
contact: eavanvalkenburg

> date: 2026-06-30
> deciders: eavanvalkenburg

Reviewer 3: nit

```suggestion
date: 2026-07-07
deciders: eavanvalkenburg
consulted: rogerbarreto
```

---

# Python protocol helpers and optional execution state

## Context and Problem Statement

Agent Framework needs to help applications expose agents and workflows over external protocols such as OpenAI
Responses, Telegram, Activity Protocol, and future transports.

FastAPI, Starlette, Azure Functions, Django, Telegram SDKs, Bot Framework SDKs, and other app frameworks already own
route registration, dependency injection, middleware, authentication, background tasks, lifecycle, and native client
calls. Agent Framework should not duplicate those surfaces unless a specific hosting environment requires it.

## Decision Drivers

- Keep the released surface small enough to explain without first teaching a channel framework.

> - Provide reusable Agent Framework run translation that works with FastAPI and other web frameworks.

Reviewer 3: Not sure why FastAPI is getting named specifically, if we mention names ideally for an abstraction I would expect 2+ named options to start with.

```suggestion
- Provide reusable Agent Framework run translation that works with all web frameworks mentioned, starting with FastAPI and Responses implementations.
```

Author: mostly because it is currently one of the most popular ones for Python, I use FastAPI and Django below in the sketch of the implementation, adding Django here as an example as well.

- Let app/framework code own route declaration, auth, middleware, native SDK clients, command handling, and background
  work.
- Keep stateful execution support explicit: session lookup/storage and workflow checkpoint lookup/storage may still need
  a small AF-owned home.

> - Avoid approving cross-channel identity and delivery semantics before their safety model is reviewed.

Reviewer 3: Not sure this is ADR related, sounds like instructions for the Agent.

```suggestion
```

Author: it related to the scope of this ADR, but we can remove.

## Considered Options

1. Create protocol-specific hosts.
2. Ship a full host/channel framework with route contribution and channel hooks.
3. Ship protocol conversion helpers plus optional execution state.

> ### Create protocol-specific hosts

Reviewer 3: I would put the numbers in the options so I don't have to come back to know which is which as well as a short description what is the actual option approach, possibly with a mermaid graph or something visual.

```suggestion
### 1. Create protocol-specific hosts

Small description of the proposal
```

Author: good point

- Good: no new shared abstraction.
- Neutral: each protocol host can evolve independently.
- Bad: every package reinvents AF input/result mapping, session-key conventions, and stateful execution helpers.

### Ship a full host/channel framework

- Good: one object can assemble routes, channels, session handling, hooks, and lifecycle callbacks.
- Good: app code using the supported host shape can be short.
- Bad: the framework owns concerns already handled by web frameworks, protocol SDKs and/or other services.
- Bad: users must understand `Channel`, contribution, hook, and host-dispatch concepts before they can see how a request
  becomes `agent.run(...)`.

> - Bad: the abstraction is hard to reuse outside the chosen ASGI shape.

Reviewer 3: What is ASGI?

```suggestion
- Bad: the abstraction is hard to reuse outside the chosen <ASGI> shape.
```

Author: Asynchronous Server Gateway Interface, a standard way of defining webservers for python (as opposed to WSGI which is non-async, given the nature of agents, async is def preferred here)

Author: made the language clearer now

### Ship protocol helpers plus optional execution state

- Good: protocol packages provide the Agent Framework run value directly: `<protocol>_to_run(...)` and
  `<protocol>_from_run(...)` style helpers.
- Good: apps keep native FastAPI, Starlette, Azure Functions, Django, Bot Framework, or Telegram SDK code.
- Good: helper functions can be tested without an ASGI app or host pipeline.
- Good: small state objects can still own target-coupled state: `AgentState` pairs an agent target with a `SessionStore`,
  and `WorkflowState` resolves a workflow target while reusing the existing `CheckpointStorage` abstraction.
- Good: provides maximum configurability in handling input and outputs (outside of the conversions)
- Bad: building a first iteration of a new Host is more verbose.
- Bad: samples show more explicit route/client code than a fully assembled channel host.

## Decision Outcome

> Chosen option: **protocol helpers plus optional execution state**.

Reviewer 3: nit.

```suggestion
Chosen option: **3. Ship protocol helpers plus optional execution state**.
```

Protocol packages own:

- parsing protocol-native input into Agent Framework run input and options;
- rendering `AgentResponse`, `AgentResponseUpdate`, workflow results, or workflow updates back into protocol-native
  response/event payloads;
- protocol-specific isolation/session id helper functions when useful, such as `telegram_session_id(update)`;
- protocol-specific typing/update event helpers where the protocol has a native concept.

Application or web-framework code owns:

- HTTP route declaration and route grouping;
- dependency injection;
- authentication and authorization;
- middleware;
- background tasks and webhook acknowledgement policy;
- native protocol SDK clients and outbound calls;
- command registration and command dispatch;
- request/response status codes and framework-specific error handling;
- choosing the isolation/session id source for the current deployment and route.

The optional execution-state helpers, if provided, are limited to shared execution state:

- `AgentState`: one `SupportsAgentRun`-compatible target plus a `SessionStore`;

> - `WorkflowState`: one `Workflow`, `WorkflowBuilder`-shaped builder, orchestration builder, or workflow factory;

Reviewer: Could a concrete `Workflow` target share runner state across callers in a persistent host? `Workflow.run` preserves state on the instance between calls, so a singleton `WorkflowState(Workflow)` can carry one authorized session's executor/shared state into another unless every run restores an isolated checkpoint or builds a fresh workflow. Should this constrain concrete `Workflow` targets to single-session use and recommend a builder/factory for multi-session hosting?

Author: so, later on I mention that this is up to the developer and depends on many things, persistent servers require a different setup, likely with a workflow factory/builder rather then a single concrete workflow, but if you are running a transient server (like a azfunc) that distinction is somewhat less relevant, but I wanted to give the developer this choice, rather then making a choice and then having to build in parameters to control this behavior.

Author: and of course our samples need to be very explicit about this.

hides the protocol-specific contract behind a second abstraction.

Protocol packages should consider these helper families. This table is a set of examples, not a required protocol or
checklist. Not every protocol needs every helper, but when a protocol has the concept the naming should stay consistent:

| Helper family | Shape | Purpose |
| --- | --- | --- |
| Run conversion | `<protocol>_to_run(...)` | Convert one protocol-native call/update/request into `Agent.run` or `Workflow.run` values. |
| Final rendering | `<protocol>_from_run(...)` | Convert a final `AgentResponse` / workflow result into protocol-native response payloads or operations. |
| Stream rendering | `<protocol>_from_streaming_run(...)` | Convert `ResponseStream` / workflow updates into protocol-native events or operations. |
| Session id extraction | `<protocol>_session_id(...)` | Extract the protocol's natural continuation/partition key from the call, if present. |
| Command/action parsing | `<protocol>_command(...)` | Parse a protocol-native command/action/operation name without deciding app policy. |

Examples:

> - `responses_to_run(...)`, `responses_from_run(...)`, `responses_from_streaming_run(...)`,
>   `responses_session_id(...)`;
> - `telegram_to_run(...)`, `telegram_from_run(...)`, `telegram_from_streaming_run(...)`,
>   `telegram_session_id(...)`, `telegram_command(...)`;
> - `activity_to_run(...)`, `activity_from_run(...)`, `activity_session_id(...)`, `activity_command(...)`;
> - `discord_to_run(...)`, `discord_from_run(...)`, `discord_session_id(...)`, `discord_command(...)`.
> 

Reviewer 3: Can we have a more complete example exercising one provider like "Telegram" how would that be used, like a small more contextualized code snippet.

```suggestion
- `responses_to_run(...)`, `responses_from_run(...)`, `responses_from_streaming_run(...)`,
  `responses_session_id(...)`;
- `telegram_to_run(...)`, `telegram_from_run(...)`, `telegram_from_streaming_run(...)`,
  `telegram_session_id(...)`, `telegram_command(...)`;
- `activity_to_run(...)`, `activity_from_run(...)`, `activity_session_id(...)`, `activity_command(...)`;
- `discord_to_run(...)`, `discord_from_run(...)`, `discord_session_id(...)`, `discord_command(...)`.

Real use-case examples including more pience of code...
```

Author: that is in the appendix, wanted to keep a more complete sample together that shows how they are all used, without introducing it here.

command activity, or A2A cancellation/request action may parse through a command/action helper, but the route or SDK
handler decides whether that command clears a session, cancels a task, calls an agent, or is ignored.

Additional helper functions can be protocol-specific when the concept is not broadly shared. Examples include
`telegram_chat_id(...)`, `telegram_callback_query_id(...)`, `telegram_media_file_id(...)`,
`discord_interaction_id(...)`, `a2a_task_id(...)`, `a2a_context_id(...)`, and MCP tool/prompt/resource helpers. These
helpers should still stay side-effect-free: they extract, normalize, or describe protocol data, while app/native SDK code
performs acknowledgements, sends/edits messages, resolves protected file URLs, applies rate limits, and registers
handlers.

### Security responsibilities for application builders

The application builder owns the trust boundary. Protocol helper packages can parse native payloads and expose candidate
ids or operations, but they do not authenticate callers, authorize access to state, or decide which side effects are
allowed.

> Application code that uses these helpers must:

Reviewer 3: We might also capitalize the individual items.

```suggestion
Application code that uses these helpers are responsible for:
```

  state for it;
- bind externally supplied ids to the authenticated user, tenant, workspace, installation, or chat context before using
  them as `SessionStore` keys or checkpoint cursor keys;
- treat `<protocol>_session_id(...)` results as untrusted candidate keys until that ownership check has passed;
- keep platform-provided isolation helpers fail-closed outside their trusted hosting environment;
- authorize command/action effects such as reset, cancel, approve, submit, or tool invocation after parsing them;
- opt in explicitly before resolving protected media/resource/file URLs and passing them to a remote model provider;
- persist post-run session or checkpoint state only after `agent.run(...)`, `workflow.run(...)`, or stream finalization has
  updated that state.

For Foundry specifically, helpers may read values established by Foundry hosting middleware, but must not treat raw
request headers as trusted Foundry isolation when the app is running outside Foundry. Implementations must test that
non-Foundry requests do not accept spoofable isolation headers as platform-provided keys.

> For durable workflow checkpointing, the checkpoint boundary must be at least as specific as the authorized session/tenant

Reviewer 3: Durable could have many meanings here, is maybe Long running could better describe it.

```suggestion
For long-running workflow checkpointing, the checkpoint boundary must be at least as specific as the authorized session/tenant
```

  durable state for multi-replica continuity;
- transient apps must not rely on in-memory `SessionStore` state between calls and need a durable session store or a
  service-owned continuation id;
- workflow hosts must choose an explicit `CheckpointStorage` and, when they need per-session resume, a durable
  `session_id -> checkpoint_id` cursor because in-process workflow state and in-memory checkpoint cursors do not survive
  transient execution.

A `SessionStore` stores `session_id -> AgentSession`, but it does not create sessions. `AgentState` resolves the agent
target and creates the session on first use:

For agent targets:

```python
session = await state.get_or_create_session(session_id)
target = await state.get_target()
result = await target.run(messages, session=session, options=options)
```

If the protocol mints a new continuation id as part of the response being created (for example, OpenAI Responses
`resp_*` ids), store the **post-run** session explicitly under that new id:

```python
session = await state.get_or_create_session(previous_response_id)
target = await state.get_target()
result = await target.run(messages, session=session, options=options)

> await state.set_session(response_id, session)

Reviewer 3: Not sure why is the caller responsibility to update the session and keep track of the new post-run id, AFAIK the session instance object already carries this responsibility.

Author: because this is specific to the way you want to run your server, if you do not want persistence, then you wouldn't call this method.

state, but it should not wrap or replace the existing `CheckpointStorage` abstraction. Apps should pass the actual
`CheckpointStorage` they want the workflow to use. If an app needs per-session resume, it can keep a small cursor from
authorized `session_id` to `checkpoint_id` (or an equivalent store-specific resume token).

Workflow runs do not currently emit a checkpoint id on `WorkflowRunResult` or normal workflow events by default. The
runner receives checkpoint ids internally from `CheckpointStorage.save(...)`. App/state code that owns the storage can
observe the latest id by querying the storage after a run, for example
`await storage.get_latest(workflow_name=target.name)`.

For workflow targets, app code adapts the protocol helper output into the workflow's expected input and invokes the
workflow through the state object's target:

```python
# session_id must already be authenticated and authorized for this caller
target = await state.get_target()
result = await target.run(message=workflow_input, checkpoint_storage=checkpoint_storage)
latest = await checkpoint_storage.get_latest(workflow_name=target.name)

> if latest is not None:

Reviewer: Could this race when `checkpoint_storage` is shared across sessions? `get_latest(workflow_name=target.name)` is global to the workflow name in the current `CheckpointStorage` protocol, so session B can save a later checkpoint between session A's run and this lookup, causing A's cursor to resume B's state. 

The ownership checks above do not mitigate a checkpoint id selected after authorization; should the sketch require per-session scoped storage here or capture the checkpoint id from this run's `save(...)` path instead of querying global latest?

Author: yes, see my note on the other comment, and it might be useful to see if we need to add something to checkpoint storage to account for this

else:
    result = await target.run(checkpoint_id=checkpoint_id, checkpoint_storage=checkpoint_storage)
latest = await checkpoint_storage.get_latest(workflow_name=target.name)
if latest is not None:
    await checkpoint_cursor_store.set(session_id, latest.checkpoint_id)
```

`workflow.run(...)` writes checkpoints to the provided storage, so storage selection must be explicit at the route layer.
Protocol helper packages should not own checkpoint layout, route lifecycle, or durable execution.

## Non-goals for v1

The following remain outside the v1 contract:

- cross-channel identity linking (`IdentityLinker`, `local_identity_link`, or `agent-framework-hosting-entra`);
- identity allowlists or authorization policy (`IdentityAllowlist`, `AuthPolicy`);
- response routing beyond the originating protocol (`ResponseTarget`, active channel, specific linked channel,
  `all_linked`);
- push or payload codecs (`ChannelPush`, `ChannelPushCodec`);
- background/continuation delivery;
- durable task runners (`DurableTaskRunner`, `InProcessTaskRunner`);
- retry/replay policy (`RetryPolicy`);
- fan-out, multicast, or all-linked delivery;
- confidentiality tiers and `LinkPolicy`;

> - a host-level multi-agent router.

Reviewer 3: Will the protocol helper option be able to cover the proposed items in the list in the future?

Author: did an update to rewrite these and noted how they are impacted by the new design

where native framework code remains in charge.

### Optional execution state

`AgentState` and `WorkflowState` stay small: they are target-specific state holders, not app hosts.

```python
from typing import Protocol

from agent_framework import AgentSession, SupportsAgentRun, Workflow

class SupportsBuild(Protocol):
    def build(self) -> Workflow: ...

class SessionStore:
    async def get(self, session_id: str) -> AgentSession | None: ...
    async def set(self, session_id: str, session: AgentSession) -> None: ...
    async def delete(self, session_id: str) -> None: ...

class CheckpointCursorStore:
    async def get(self, session_id: str) -> str | None: ...
    async def set(self, session_id: str, checkpoint_id: str) -> None: ...
    async def delete(self, session_id: str) -> None: ...

class AgentState:
    def __init__(self, target: SupportsAgentRun, *, session_store: SessionStore | None = None) -> None: ...
    async def get_target(self) -> SupportsAgentRun: ...
    async def get_or_create_session(self, session_id: str) -> AgentSession: ...
    async def set_session(self, session_id: str, session: AgentSession) -> None: ...

> class WorkflowState:

Reviewer 3: Suggest also adding the changes to the PR, so any extra findings including use-cases, code paths, limitations, are reflected into the ADR observation.
